docs/advanced_documentation/algorithms/lu-solver.md

```diff
@@ -0,0 +1,714 @@
+at (0-based) indices $(i,j)$, where $i,j = 0..(N-1)$.
+
+1. Initialize the permutations $P$ and $Q$ to the identity permutation.
+2. Initialize fill-in elements to $0$.
+3. Loop over all rows: $p = 0..(N-1)$:
+   1. Set the remaining matrix: $M_p \gets M\left[p:N,p:N\right]$ with size $N_p\times N_p$.
+   2. Find largest element $M_p\left[i_p,j_p\right]$ in $M_p$ by magnitude. This is the pivot
+      element.
+   3. If the magnitude of the pivot element is too small:
+      1. If pivot perturbation is enabled, then:
+         1. Apply [pivot perturbation](#pivot-perturbation).
+         2. Proceed.
+      2. Else, if the matrix is singular (pivot element is identically $0$), then:
+         1. Raise a `SparseMatrixError`.
+      3. Else:
+         1. Proceed.
+   4. Else:
+      1. Proceed.
+   5. Swap the first and pivot row and column of $M_p$, so that the pivot element is in the
+      top-left corner of the remaining matrix:
+      1. $M_p\left[0,0:N_p\right] \leftrightarrow M\left[i_p,0:N_p\right]$
+      2. $M_p\left[0:N_p,0\right] \leftrightarrow M\left[0:N_p,j_p\right]$
+   6. Apply Gaussian elimination for the current pivot element:
+      1. $M_p\left[0,0:N_p\right] \gets \frac{1}{M_p[0,0]}M_p\left[0,0:N_p\right]$
+      2. $M_p\left[1:N_p,0:N_p\right] \gets M_p\left[1:N_p,0:N_p\right] - M_p\left[1:N_p,0\right] \otimes M_p\left[0,0:N_p\right]$
+   7. Accumulate the permutation matrices:
+      1. In $P$: swap $p \leftrightarrow p + i_p$
+      2. In $Q$: swap $p \leftrightarrow p + j_p$
+   8. Continue with the next $p$ to factorize the the bottom-right block.
+
+$L$ is now the matrix composed of ones on the diagonal and the bottom-left off-diagonal triangle of
+$M$ and zeros in the upper-right off-diagonal triangle. $Q$ is the matrix composed of the
+upper-right triangle of $M$ including the diagonal elements and zeros in the lower-left off-diagonal
+triangle.
+
+```{note}
+In the actual implementation, we use a slightly different order of operations: instead of raising
+the `SparseMatrixError` immediately, we break from the loop and throw after that. This does not
+change the functional behavior.
+```
+
+### Block-sparse LU-factorization
+
+The LU-factorization process for block-sparse matrices is similar to that for
+[dense matrices](#dense-lu-factorization), but in this case, $m_p$ is a block element, and
+$\vec{q}_p$, $\vec{r}_p^T$ and $\hat{M}_p$ consist of block elements as well. Notice that the
+inverse $m_p^{-1}$ can be calculated from is LU-decomposition, which can be obtained from the
+[dense LU-factorization process](#dense-lu-factorization-process).
+
+#### Block-sparse LU-factorization process
+
+The Gaussian elimination process itself is as usual. Completely analogously to and following the
+same conventions as [before](#dense-lu-factorization-process), let
+$\underline{M}_p\equiv\begin{bmatrix}\underline{m}_p && \vec{\underline{r}}_p^T \\ \vec{\underline{q}}_p && \hat{\underline{M}}_p\end{bmatrix}$
```

Comment:
1. Why do you use underscore? Underscore in EE means specifically complex numbers.
2. Same comment as above. Using 4 blocks instead of 9 blocks is confusing for the readers. Reader might think this is only applicable for the first pivot.

Reply:
I think i now understand what you mean. I'm describing the process at some point in the iteration during treatment of pivot element $p$.

I can indeed extend it, but the $p-1$ row and column do not contribute to this part of the process. What do you recommend?

Reply:
The $0 \to p-1`$ does contribute in the process. If you permute columns/rows to get the max pivot for element $p$. The previous columns and rows have to be permuted as well.

Also, for the reader, include previous rows/columns shows that the factorization is in the middle of the process instead of just the beginning.

Reply:
@TonyXiang8787 can you please double-check whether this is corrrectly resolved in the latest version?